### PR TITLE
add example of parallel processing California housing data

### PR DESCRIPTION
**Pull request by siliang-j-1225**
add example of parallel processing California housing data

### CODE REVIEW

**Review by daisukei777 — DISMISSED** [review]

**Review by aykhara — DISMISSED** [review]

File: samples/parallel-processing-california-housing-data/.flake8 [hunk 1]
```diff
@@ -0,0 +1,3 @@
+[flake8]
+ignore = E501,E741,E402
+exclude = .git, __pycache__, docs, ml_model/utils, ml_model/conf, ml_model/steps
```
Comment (h2floh) [inline samples/parallel-processing-california-housing-data/.flake8]: What is the reason to exclude ml_model subfolders? Just curious, nothing I would block for.

File: samples/parallel-processing-california-housing-data/README.md [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+This sample takes input files from Azure Machine Learning Datastore which mapped to Azure Blob Storage.
+
+#### Create blob container and folders
+
+1. Go to Azure Storage Account and create new container named 'azureml'
+
+#### Upload input data
+
+1. change directory to `data` and run `create_sample_data.py` to create sample data.
```
Comment (h2floh) [inline samples/parallel-processing-california-housing-data/README.md]:
```suggestion
1. Change directory to `data` and run `create_sample_data.py` to create sample data.
```

File: samples/parallel-processing-california-housing-data/README.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+1. Use the VSCode [dev container](./.devcontainer), or install Anaconda or Mini Conda and create a Conda envrionment by running [local_install_requirements.sh](./local_development/local_install_requirements.sh).
+
+1. In VSCode, open the root folder of this sample, select the Conda environment created above as the Python interpretor.
+
+1. Publish and run Azure ML pipelines. 
+    - To publish and run Azure ML pipelines, run:
+        ```shell
+        # publish the Azure ML pipeline
+        python -m ml_service.pipelines.build_pipeline
```
Comment (h2floh) [inline samples/parallel-processing-california-housing-data/README.md:137]: Ok, cool so we have the AML pipeline created. Now wouldn't it be great to explain also how to run the pipeline? In order to get some action :)

File: samples/parallel-processing-california-housing-data/README.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
+# Overview
+
+**What does this sample demonstrate:**
+- A simple example of how to use `ParallelRunScript` to process data in parallel.
```
Comment (h2floh) [inline samples/parallel-processing-california-housing-data/README.md]: The whole sample is to demonstrate how efficient the parallel processing step is in these kind of scenarios. Would be great if you could explain here what is done in parallel and that we experienced up to x % time reduction.

File: samples/parallel-processing-california-housing-data/local_development/.env.example [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+PIPELINE_NAME = "california_housing_pipeline"
+
+# Path
+INPUT_DIR="input"
+SOURCES_DIR_TRAIN = "ml_model"
+PREPARATION_STEP_SCRIPT_PATH = "preparation/preparation_aml.py"
+EXTRACTION_STEP_SCRIPT_PATH = "extraction/extraction_aml.py"
+TRAINING_STEP_SCRIPT_PATH = "training/training_aml.py"
+
+# Azure ML Parallel Run Config for Language Model Pipeline
```
Comment (h2floh) [inline samples/parallel-processing-california-housing-data/local_development/.env.example:32]: You explained the parameters for the parallel step verbally very nice, would be great to have a section in the README where you explain the options in a bit more detail 👍

File: samples/parallel-processing-california-housing-data/ml_model/preparation/preparation_aml.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
+"""Preparation wrapper: script to be executed when running Utterance Step."""
```
Comment (h2floh) [inline samples/parallel-processing-california-housing-data/ml_model/preparation/preparation_aml.py]:
```suggestion
"""Preparation wrapper: script to be executed when running Preparation Step."""
```

File: samples/parallel-processing-california-housing-data/ml_model/training/training_aml.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
+"""Training wrapper: script to be executed when running Utterance Step."""
```
Comment (h2floh) [inline samples/parallel-processing-california-housing-data/ml_model/training/training_aml.py]:
```suggestion
"""Training wrapper: script to be executed when running Training Step."""
```

File: samples/parallel-processing-california-housing-data/local_development/dev_dependencies.yml [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+  - pip:
+      - azureml-sdk
+      - azure-storage-file-datalake
+      - black
+      - flake8
+      - flake8_formatter_junit_xml
+      - isort
+      - pandas
+      - pydocstyle
+      - pytest
```
Comment (h2floh) [inline samples/parallel-processing-california-housing-data/local_development/dev_dependencies.yml]: Just a question: as you exclude testing explicitly from the example, there wouldn't be a need for pytest* dependencies, right?

File: samples/parallel-processing-california-housing-data/README.md [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+1. In VSCode, open the root folder of this sample, select the Conda environment created above as the Python interpretor.
+
+1. Publish and run Azure ML pipelines. 
+    - To publish and run Azure ML pipelines, run:
+        ```shell
+        # publish the Azure ML pipeline
+        python -m ml_service.pipelines.build_pipeline
+        ```
+
```
Comment (h2floh) [inline samples/parallel-processing-california-housing-data/README.md]: Would be really great to have a section here explaining briefly the additional parameters for the ParallelScriptStep. I made a comment on the `.env.example` as well.

**Review by h2floh — COMMENTED** [review]
Hi @siliang-j-1225 , thanks for your contribution! Looks really awesome. I made some comments and suggestions. Hope you like them!

File: samples/parallel-processing-california-housing-data/README.md (hunk 3 above)
Comment (siliang-j-1225) [reply samples/parallel-processing-california-housing-data/README.md:137]: that's a good idea! I will add more documentation here.

File: samples/parallel-processing-california-housing-data/.flake8 (hunk 1 above)
Comment (siliang-j-1225) [reply samples/parallel-processing-california-housing-data/.flake8]: @kuimoani @kenakamu  do you know why ml_model is excluded from linter? I actually copied this setting from yesno example.

File: samples/parallel-processing-california-housing-data/README.md (hunk 3 above)
Comment (siliang-j-1225) [reply samples/parallel-processing-california-housing-data/README.md:137]: I have added simple explanation of how to access to aml to submit a new run. pls have a look : )

File: samples/parallel-processing-california-housing-data/README.md (hunk 4 above)
Comment (siliang-j-1225) [reply samples/parallel-processing-california-housing-data/README.md]: updated in the new commit https://github.com/liupeirong/MLOpsManufacturing/pull/69/commits/5a6dbbf4d8bc97a67a37db14f1a99dc50637c5e2.

File: samples/parallel-processing-california-housing-data/local_development/.env.example (hunk 5 above)
Comment (siliang-j-1225) [reply samples/parallel-processing-california-housing-data/local_development/.env.example:32]: updated in the new commit https://github.com/liupeirong/MLOpsManufacturing/pull/69/commits/5a6dbbf4d8bc97a67a37db14f1a99dc50637c5e2.

File: samples/parallel-processing-california-housing-data/README.md (hunk 9 above)
Comment (siliang-j-1225) [reply samples/parallel-processing-california-housing-data/README.md]: updated in the new commit https://github.com/liupeirong/MLOpsManufacturing/pull/69/commits/5a6dbbf4d8bc97a67a37db14f1a99dc50637c5e2. I have separated the parallel run variable explanation from other variables. and also added more details for each of them.

File: samples/parallel-processing-california-housing-data/.flake8 (hunk 1 above)
Comment (h2floh) [reply samples/parallel-processing-california-housing-data/.flake8]: ah ok, in that sample there are a lot of scripts from external, which we don't want to modify so therefore the exclusion. We can remove it here.
```suggestion
exclude = .git, __pycache__, docs
```

**Review by h2floh — DISMISSED** [review]
@siliang-j-1225 thanks for your contribution! LGTM

**Review by h2floh — APPROVED** [review]

**Pull request merged**